apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/commons/AbstractSamhsaMatcherTest.java

```diff
@@ -0,0 +1,812 @@
+import gov.cms.bfd.server.war.adapters.ProcedureComponent;
+import gov.cms.bfd.server.war.utils.ReflectionTestUtils;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.mockito.ArgumentCaptor;
+
+@RunWith(Enclosed.class)
+public class AbstractSamhsaMatcherTest {
+
+  private static final IllegalStateException INVOCATION_EXCEPTION =
+      new IllegalStateException("Method was not called with expected arguments for test");
+
+  @FunctionalInterface
+  public interface SamhsaFilterMethod<T> {
+    boolean apply(AbstractSamhsaMatcher<IBaseResource> matcher, T coding);
+  }
+
+  public static class NonParameterizedTests {
+
+    @Test
+    public void shouldReturnColumnValues() {
+      List<String> expected = List.of("1", "2", "3", "4");
+      List<String> actual =
+          AbstractSamhsaMatcher.resourceCsvColumnToList("samhsa_codes_test_file.csv", "columnB");
+
+      assertEquals(expected, actual);
+    }
+
+    @Test
+    public void shouldReturnTrueForSamhsaProcedureComponent() {
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      AbstractSamhsaMatcher<IBaseResource> matcherSpy = spy(AbstractSamhsaMatcher.class);
+
+      ProcedureComponent mockComponentA = mock(ProcedureComponent.class);
+      ProcedureComponent mockComponentB = mock(ProcedureComponent.class);
+
+      doThrow(INVOCATION_EXCEPTION)
+          .when(matcherSpy)
+          .isSamhsaIcdProcedure(any(ProcedureComponent.class));
+
+      doReturn(false).when(matcherSpy).isSamhsaIcdProcedure(mockComponentA);
+
+      doReturn(true).when(matcherSpy).isSamhsaIcdProcedure(mockComponentB);
+
+      assertTrue(
+          "Samhsa procedure not correctly filtered",
+          matcherSpy.containsSamhsaIcdProcedureCode(List.of(mockComponentA, mockComponentB)));
+    }
+
+    @Test
+    public void shouldReturnFalseForNonSamhsaProcedureComponent() {
+      // unchecked - This is ok for making a mock.
+      //noinspection unchecked
+      AbstractSamhsaMatcher<IBaseResource> matcherSpy = spy(AbstractSamhsaMatcher.class);
+
+      ProcedureComponent mockComponentA = mock(ProcedureComponent.class);
+      ProcedureComponent mockComponentB = mock(ProcedureComponent.class);
+
+      doThrow(INVOCATION_EXCEPTION)
+          .when(matcherSpy)
+          .isSamhsaIcdProcedure(any(ProcedureComponent.class));
+
+      doReturn(false).when(matcherSpy).isSamhsaIcdProcedure(mockComponentA);
+
+      doReturn(false).when(matcherSpy).isSamhsaIcdProcedure(mockComponentB);
+
+      assertFalse(
+          "Non-Samhsa procedure incorrectly filtered",
+          matcherSpy.containsSamhsaIcdProcedureCode(List.of(mockComponentA, mockComponentB)));
+    }
+
```

Comment:
Sure, I'm just wondering if that is a bit over-repetitive?  The logic is that a collection of procedures may contain a samhsa code.  Testing for 1, 2 and 3 components seems a bit arbitrary, and we obviously can't test for 1...n components.  It's been my experience that if you try to test every possible scenario, you end up with very unmanageable tests, and you still wouldn't have tested every possible scenario.

Reply:
I think just testing for 1 and then a single multiple group is sufficient; I recommend testing this way due to past experiences with loop logic sometimes being incorrect/error-prone when iterating over multiple items, which can make a test that tests one item pass and a test that checks 4 items fail. Certainly wouldnt advocate for testing every possible combination :) 

I think if the list of items is set up in a way that tests the logic and the item to be caught is in the middle of a list, its a bit safer (wont be caught by first or last position bugs) and we could probably assume that it'd work for a single item list as well. If we did avoid the single-item list test, it might be worth documenting the assumption in the test description of the multi-test that we expect this to work for a single item list as well.